### ISSUE TITLE
1.0 API discussion

### ISSUE THREAD

**Issue by stephank** [AngryBytes/rawacme-node#2]
Pinging @pkunzexternal and @doublerebel (via npm dependents)

Cool to see some use for this package! 😀 If there's any issues you're having, I'd love to hear about them. Activity in this repo may be low, but we've been running with this in production for quite some time now.

I'm opening up this issue to talk about any API changes you'd like to see. A next version will likely coincide with the implementation of the new ACME standard (#1).

Two things I have in mind, to get started:

 - Callbacks optional, returning promises when omitted. (Need to figure out the convention here, but I believe libraries should just use the global `Promise`? And then, users can set `global.Promise` to a custom implementation like Bluebird if they want.)

 - Change the way we query the directory. I feel like the whole thing should be optional, to allow caching and the like. How about changing the flow to:

```js
const RawAcme = require('rawacme')

// Now a regular constructor, and no `url` arg.
const client = new RawAcme({ publicKey: '...', privateKey: '...' })
// Fetches the directory, adds methods to the client object, and returns the JSON body.
const directory = await client.importDirectory(RawAcme.LETSENCRYPT_STAGING)
// We can now use directory methods as before.
const res = await client.newReg({ contact: ['mailto:john@example.com'] })

// User may cache the directory JSON body, and reuse it on a new client instance.
const client2 = new RawAcme({ publicKey: '...', privateKey: '...' })
await client2.importDirectory(directory)  
```

(And the various constants and helpers we now export would become static properties of `RawAcme`, so we just have one export.)

Thoughts?

**Comment by philkunz** [comment]
Hi, interesting. :) I'll take a look at it and report back. btw. @philkunz is the correct account, @pkunzexternal is just for keeping track of stuff when the company does stuff as contractor for "external" companies. ;)

**Comment by doublerebel** [comment]
I agree, let's use the global Promise.  Global promise works very well as I move to async/await now that we've hit Node 8.  I do like having callbacks for onboarding simplicity until async/await is more prevalent.

I'm also a fan of Instances over Singletons, so in that sense supporting multiple directories is a no-brainer.

Otherwise the library has been very easy to work with, but I'll give it a look to see if there's something more I could from the interface.   Thanks!

**Comment by philkunz** [comment]
What are the thoughts on moving to TypeScript? It would make the API much easier to work with + we can use async/await without breaking support for node v6.

**Comment by stephank** [comment]
This issue is rather old, but we're still using this library. So I did some cowboy coding on branch `acmev2` to get our stack up to date. This is also published as the `beta` tag in npm, currently version `v0.3.0-beta.1`.

This new version only supports ACMEv2, is converted to TypeScript, and has a bunch of breaking API changes. Hopefully the readme and inline comments are descriptive? (But the ACME parts outside the scope of this library have also changed quite a bit.)

I've been testing against a local Pebble server, and have also updated our internal project to hopefully eliminate any rough edges. However, I've not yet tested against Let's Encrypt, or taken this to production yet. Those will be the next steps.

**Comment by stephank** [comment]
We've now been running this in production for a couple months, and all looks good. I've released v0.3.0 with only small doc fixes.

I'm personally not too worried about needing a 1.0, so will just close this ticket now.

Thanks for all the feedback! :)
